File: SortVision/public/code/insertion/rust/insertionSort.rs
// Insertion Sort (in-place)
// Time: Best O(n), Avg/Worst O(n^2), Space: O(1)
pub fn insertion_sort(arr: &mut [i32]) {
    for i in 1..arr.len() {
        let key = arr[i];
        let mut j = i;

        while j > 0 && arr[j - 1] > key {
            arr[j] = arr[j - 1];
            j -= 1;
        }
        arr[j] = key;
    }
}

fn main() {
    let mut arr = [64, 25, 12, 22, 11];
    insertion_sort(&mut arr);
    println!("{:?}", arr);
}

File: SortVision/public/code/insertion/python/insertionSort.py
"""Insertion Sort (in-place).

Time: Best O(n), Avg/Worst O(n^2)
Space: O(1)
"""

def insertion_sort(arr: list[int]) -> list[int]:
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1

        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key
    return arr


if __name__ == "__main__":
    arr = [64, 25, 12, 22, 11]
    print(insertion_sort(arr))
